# Exploratory Data Analysis (EDA): ClinVar Variant Classifications

### Objective:
The goal of this analysis is to explore the dataset and understand the underlying patterns that differentiate between **Concordant (Class 0)** and **Conflicting (Class 1)** variant interpretations.

### Key Steps:
1. **Data Loading:** Importing the dataset and initial inspection.
3. **Exploratory Analysis:** Explore data.

### Importing Libraries & Loading Data (Code)

In [27]:
import pandas as pd
pd.set_option('display.max_columns',None)
import numpy as np

In [28]:
# Load the dataset
df = pd.read_csv('/Users/rghdmajd/PycharmProjects/ClinVar_Project/data/raw/clinvar_conflicting.csv',low_memory=False)

# Display the first 5 rows to understand the structure
df.head()

,CHROM,POS,REF,ALT,AF_ESP,AF_EXAC,AF_TGP,CLNDISDB,CLNDISDBINCL,CLNDN,CLNDNINCL,CLNHGVS,CLNSIGINCL,CLNVC,CLNVI,MC,ORIGIN,SSR,CLASS,Allele,Consequence,IMPACT,SYMBOL,Feature_type,Feature,BIOTYPE,EXON,INTRON,cDNA_position,CDS_position,Protein_position,Amino_acids,Codons,DISTANCE,STRAND,BAM_EDIT,SIFT,PolyPhen,MOTIF_NAME,MOTIF_POS,HIGH_INF_POS,MOTIF_SCORE_CHANGE,LoFtool,CADD_PHRED,CADD_RAW,BLOSUM62
0,1,1168180,G,C,0.0771,0.10020,0.1066,MedGen:CN169374,NaN,not_specified,NaN,NC_000001.10:g.1168180G>C,NaN,single_nucleotide_variant,UniProtKB_(protein):Q96L58#VAR_059317,SO:0001583|missense_variant,1,NaN,0,C,missense_variant,MODERATE,B3GALT6,Transcript,NM_080605.3,protein_coding,1/1,NaN,552,522,174,E/D,gaG/gaC,NaN,1.0,NaN,tolerated,benign,NaN,NaN,NaN,NaN,NaN,1.053,-0.208682,2.0
1,1,1470752,G,A,0.0000,0.00000,0.0000,"MedGen:C1843891,OMIM:607454,Orphanet:ORPHA9877...",NaN,Spinocerebellar_ataxia_21|not_provided,NaN,NC_000001.10:g.1470752G>A,NaN,single_nucleotide_variant,OMIM_Allelic_Variant:616101.0001|UniProtKB_(pr...,SO:0001583|missense_variant,1,NaN,0,A,missense_variant,MODERATE,TMEM240,Transcript,NM_001114748.1,protein_coding,4/4,NaN,523,509,170,P/L,cCg/cTg,NaN,-1.0,OK,deleterious_low_confidence,benign,NaN,NaN,NaN,NaN,NaN,31.000,6.517838,-3.0
2,1,1737942,A,G,0.0000,0.00001,0.0000,"Human_Phenotype_Ontology:HP:0000486,MedGen:C00...",NaN,Strabismus|Nystagmus|Hypothyroidism|Intellectu...,NaN,NC_000001.10:g.1737942A>G,NaN,single_nucleotide_variant,OMIM_Allelic_Variant:139380.0002|UniProtKB_(pr...,"SO:0001583|missense_variant,SO:0001623|5_prime...",35,NaN,1,G,missense_variant,MODERATE,GNB1,Transcript,NM_002074.4,protein_coding,6/12,NaN,632,239,80,I/T,aTc/aCc,NaN,-1.0,OK,deleterious,probably_damaging,NaN,NaN,NaN,NaN,NaN,28.100,6.061752,-1.0
3,1,2160305,G,A,0.0000,0.00000,0.0000,"MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002...",NaN,Shprintzen-Goldberg_syndrome|not_provided,NaN,NC_000001.10:g.2160305G>A,NaN,single_nucleotide_variant,OMIM_Allelic_Variant:164780.0004|UniProtKB_(pr...,SO:0001583|missense_variant,33,NaN,0,A,missense_variant,MODERATE,SKI,Transcript,XM_005244775.1,protein_coding,1/7,NaN,132,100,34,G/S,Ggc/Agc,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,22.500,3.114491,NaN
4,1,2160305,G,T,0.0000,0.00000,0.0000,"MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002",NaN,Shprintzen-Goldberg_syndrome,NaN,NC_000001.10:g.2160305G>T,NaN,single_nucleotide_variant,OMIM_Allelic_Variant:164780.0005|UniProtKB_(pr...,SO:0001583|missense_variant,33,NaN,0,T,missense_variant,MODERATE,SKI,Transcript,XM_005244775.1,protein_coding,1/7,NaN,132,100,34,G/C,Ggc/Tgc,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,24.700,4.766224,-3.0


## Data Overview

In [29]:
# Check basic info and column types
print("Dataset Shape:", df.shape)
print("-" * 30)
df.info()

Dataset Shape: (65188, 46)
------------------------------
<class 'pandas.DataFrame'>
RangeIndex: 65188 entries, 0 to 65187
Data columns (total 46 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   CHROM               65188 non-null  str    
 1   POS                 65188 non-null  int64  
 2   REF                 65188 non-null  str    
 3   ALT                 65188 non-null  str    
 4   AF_ESP              65188 non-null  float64
 5   AF_EXAC             65188 non-null  float64
 6   AF_TGP              65188 non-null  float64
 7   CLNDISDB            65188 non-null  str    
 8   CLNDISDBINCL        167 non-null    str    
 9   CLNDN               65188 non-null  str    
 10  CLNDNINCL           167 non-null    str    
 11  CLNHGVS             65188 non-null  str    
 12  CLNSIGINCL          167 non-null    str    
 13  CLNVC               65188 non-null  str    
 14  CLNVI               27659 non-null  str    
 15  MC    

In [ ]:
# Display dataset columns and data types for each column
print("Dataset column:")
print(df.columns)
print("Data types of each column:")
print(df.dtypes)

Dataset column:
Index(['CHROM', 'POS', 'REF', 'ALT', 'AF_ESP', 'AF_EXAC', 'AF_TGP', 'CLNDISDB',
       'CLNDISDBINCL', 'CLNDN', 'CLNDNINCL', 'CLNHGVS', 'CLNSIGINCL', 'CLNVC',
       'CLNVI', 'MC', 'ORIGIN', 'SSR', 'CLASS', 'Allele', 'Consequence',
       'IMPACT', 'SYMBOL', 'Feature_type', 'Feature', 'BIOTYPE', 'EXON',
       'INTRON', 'cDNA_position', 'CDS_position', 'Protein_position',
       'Amino_acids', 'Codons', 'DISTANCE', 'STRAND', 'BAM_EDIT', 'SIFT',
       'PolyPhen', 'MOTIF_NAME', 'MOTIF_POS', 'HIGH_INF_POS',
       'MOTIF_SCORE_CHANGE', 'LoFtool', 'CADD_PHRED', 'CADD_RAW', 'BLOSUM62'],
      dtype='str')
Data types of each column:
CHROM                     str
POS                     int64
REF                       str
ALT                       str
AF_ESP                float64
AF_EXAC               float64
AF_TGP                float64
CLNDISDB                  str
CLNDISDBINCL              str
CLNDN                     str
CLNDNINCL                 str
CLNHGVS           

### Statistical Overview
Using `describe()` allows us to understand the central tendency, dispersion, and shape of the dataset's numerical distributions. It is vital for detecting potential outliers or data entry errors.

In [30]:
# Display summary statistics for numerical columns
summary_stats = df.describe()

# Transpose for better readability if there are many columns
print(summary_stats.transpose())

                      count          mean           std         min  \
POS                 65188.0  7.757594e+07  5.974051e+07  961.000000   
AF_ESP              65188.0  1.451052e-02  5.779541e-02    0.000000   
AF_EXAC             65188.0  1.449218e-02  5.954210e-02    0.000000   
AF_TGP              65188.0  1.526350e-02  5.952741e-02    0.000000   
ORIGIN              65188.0  1.342486e+00  5.688772e+00    0.000000   
SSR                   130.0  2.269231e+00  4.190777e+00    1.000000   
CLASS               65188.0  2.521016e-01  4.342226e-01    0.000000   
DISTANCE              108.0  8.257315e+02  1.069363e+03    1.000000   
STRAND              65174.0 -6.659097e-03  9.999855e-01   -1.000000   
MOTIF_POS               2.0  1.000000e+00  0.000000e+00    1.000000   
MOTIF_SCORE_CHANGE      2.0 -8.000000e-02  2.404163e-02   -0.097000   
LoFtool             60975.0  3.450584e-01  3.612384e-01    0.000069   
CADD_PHRED          64096.0  1.568562e+01  1.083635e+01    0.001000   
CADD_R

### Data Types Inspection
Verifying that each column is assigned the correct data type (e.g., numeric, object/categorical) is crucial. Incorrect types can prevent the model from processing features accurately.

In [32]:
# Feature dictionary classification
feature_types = {
    "numerical": {
        "continuous": ['AF_ESP', 'AF_EXAC', 'AF_TGP', 'CADD_PHRED', 'CADD_RAW', 'BLOSUM62'],
        "discrete": ['POS', 'ORIGIN']
    },
    "categorical": {
        "nominal": ['CHROM', 'REF', 'ALT', 'IMPACT', 'SYMBOL', 'Consequence'],
        "ordinal": ['CLNDN']
    }
}
feature_types
# all_numerical = feature_types["numerical"]["continuous"] + feature_types["numerical"]["discrete"]

{'numerical': {'continuous': ['AF_ESP',
   'AF_EXAC',
   'AF_TGP',
   'CADD_PHRED',
   'CADD_RAW',
   'BLOSUM62'],
  'discrete': ['POS', 'ORIGIN']},
 'categorical': {'nominal': ['CHROM',
   'REF',
   'ALT',
   'IMPACT',
   'SYMBOL',
   'Consequence'],
  'ordinal': ['CLNDN']}}

## Data Quality Assessment
Before proceeding, we need to identify the extent of missing data. Columns with excessive missing values may not be useful for prediction and might need to be dropped.

### Missing Values

In [7]:
# Calculate percentage of missing values per column
missing_percentage = (df.isnull().sum() / len(df)) * 100
missing_df = pd.DataFrame({'Missing_Percentage': missing_percentage})

# Display columns with more than 50% missing data
print(missing_df[missing_df['Missing_Percentage'] > 50])

                    Missing_Percentage
CLNDISDBINCL                 99.743818
CLNDNINCL                    99.743818
CLNSIGINCL                   99.743818
CLNVI                        57.570412
SSR                          99.800577
INTRON                       86.495981
DISTANCE                     99.834325
BAM_EDIT                     50.958765
SIFT                         61.900963
PolyPhen                     61.962324
MOTIF_NAME                   99.996932
MOTIF_POS                    99.996932
HIGH_INF_POS                 99.996932
MOTIF_SCORE_CHANGE           99.996932
BLOSUM62                     60.739707


### Data Integrity Check: Duplicates
It is essential to check for duplicate entries in the dataset, as they can bias the machine learning model and lead to overfitting.

In [14]:
# Check for duplicate rows
duplicates_count = df.duplicated().sum()
print(f"Number of duplicate rows found: {duplicates_count}")

Number of duplicate rows found: 0


### Unique Values

In [33]:
for col in df.columns:
    print(col, df[col].nunique())

CHROM 24
POS 63115
REF 866
ALT 458
AF_ESP 2842
AF_EXAC 6667
AF_TGP 2087
CLNDISDB 9234
CLNDISDBINCL 93
CLNDN 9260
CLNDNINCL 101
CLNHGVS 65188
CLNSIGINCL 137
CLNVC 7
CLNVI 27654
MC 90
ORIGIN 31
SSR 2
CLASS 2
Allele 374
Consequence 48
IMPACT 4
SYMBOL 2328
Feature_type 2
Feature 2369
BIOTYPE 2
EXON 3264
INTRON 1929
cDNA_position 13970
CDS_position 13663
Protein_position 7339
Amino_acids 1262
Codons 2220
DISTANCE 96
STRAND 2
BAM_EDIT 2
SIFT 4
PolyPhen 4
MOTIF_NAME 2
MOTIF_POS 1
HIGH_INF_POS 1
MOTIF_SCORE_CHANGE 2
LoFtool 1195
CADD_PHRED 9324
CADD_RAW 63803
BLOSUM62 6


---
## Key Insights & Summary

1. Data Quality & Structure
The dataset contains 65,188 entries and 46 features. A significant data quality issue was identified, where several columns (e.g., `MOTIF_NAME`, `SSR`, `DISTANCE`, `CLNDISDBINCL`) exhibit missing values exceeding 80%. These columns are considered noisy and will be removed during the data cleaning phase to improve model performance.

---